#!usr/bin/python3

import os
import sys
import argparse

parser = argparse.ArgumentParser(description="Script to create the input_metadata.tsv file to run fisher.py")
parser.add_argument("-p", "--path", help="space separated paths to the directories containing the proteome files generated by compleasm.py")
parser.add_argument("-o", "--out", help="output path for the input_metadata.tsv file")
paths_to_compleasm_output = parser.parse_args().path.split(" ")
output_path = parser.parse_args().out

# loop to generate and print the approriate line of text
def main():
    for file_path in paths_to_compleasm_output:
        # get the location of the proteome file
        file_path = '/' + '/'.join(file_path.split("/")[1:-1])
        location = file_path.rstrip("translated_protein.fasta")
        file_name = "translated_protein.fasta"
        unique_id = file_path.split("/")[-2]
        # remove any underscores, at symbols, double dots, white space, or asterisks
        unique_id = unique_id.replace("_", "")
        unique_id = unique_id.replace("@", "")
        unique_id = unique_id.replace("..", "")
        unique_id = unique_id.replace(" ", "")
        unique_id = unique_id.replace("*", "")
        higher_taxonomy = "Amoebozoa"
        lower_taxonomy = "Discosea"
        blast_seed = "none"
        long_name = "Unknownus nameus"
        data_type = "MAG"
        source = "SequencerProbably"

        out_line = "\t".join([location, file_name, unique_id, higher_taxonomy, lower_taxonomy, blast_seed, long_name, data_type, source])
        
        print(out_line)
        with open(output_path, "a") as out_file:
            out_file.write(out_line + "\n")

if __name__ == "__main__":
    main()
