#!usr/bin/python3

import os
import sys
import pandas as pd
import subprocess

# Script to create the input_metadata.tsv file to
# run fisher.py

# Input:
# 1. List space seperated paths to the directories containing the proteome
#    files generated by compleasm.py
#
# Output:
# 1. tab-delimited file containing the following
#    columns:
#    a. Location of the proteome file
#    b. File Name
#    c. Unique ID
#        Unique IDs cannot contain underscores “_”,
#        at symbols “@”, double dots “..”, white
#        space, or asterisk “*” 
#    d. Higher Taxonomy - always "Amoebozoa"
#    e. Lower Taxonomy - always "Discosea"
#    f. Blast Seed - always "none"
#    g. Long Name - always "Janus doeus"
#    h. Data Type - always "MAG"
#    i. Source - always "SequencerProbably"

# Usage:
# python3 fishing_meta.py <path_to_compleasm_output_dir>

# get the path to the compleasm output directory
paths_to_compleasm_output = sys.argv[1:]

# loop to generate and print the approriate line of text

for file_path in paths_to_compleasm_output:
    # get the location of the proteome file
    file_path = '/'.join(file_path.split("/")[1:-1])
    location = file_path.rstrip("translated_protein.fasta")
    file_name = "translated_protein.fasta"
    unique_id = file_path.split("/")[-2]
    # remove any underscores, at symbols, double dots, white space, or asterisks
    unique_id = unique_id.replace("_", "")
    unique_id = unique_id.replace("@", "")
    unique_id = unique_id.replace("..", "")
    unique_id = unique_id.replace(" ", "")
    unique_id = unique_id.replace("*", "")
    higher_taxonomy = "Amoebozoa"
    lower_taxonomy = "Discosea"
    blast_seed = "none"
    long_name = "Unknownus nameus"
    data_type = "MAG"
    source = "SequencerProbably"

    out_line = "\t".join([location, file_name, unique_id, higher_taxonomy, lower_taxonomy, blast_seed, long_name, data_type, source])
    print(out_line)

